(maint) Remove the visitor pattern from Puppetfile usage

justinstoller opened the pull request:
Please add all notable changes to the "Unreleased" section of the CHANGELOG in the format:
```
- (JIRA ticket) Summary of changes. [Issue or PR #](link to issue or PR)
```

justinstoller committed:
(maint) Provide Visitorless means of syncing Puppetfile

Previously to sync the content of a Puppetfile a caller had to implement
the Visitor pattern. This commit provides a `sync` method on the
Puppetfile object that does not require a visitor. It also refactors the
ContentSynchronizer to offer `concurrent_sync` and `serial_sync`
versions of its `concurrent_visit` and `serial_visit` methods.

justinstoller committed:
(maint) Remove Visitor from `puppetfile install` action

justinstoller committed:
(maint) Move deploy module to use Environment#deploy rather than the visitor pattern

justinstoller committed:
(maint) Move deploy environment to use Environment#deploy rather than the visitor pattern

justinstoller commented on spec/unit/action/deploy/environment_spec.rb:
```diff
@@ -206,21 +206,13 @@
       end
     end
 
-    describe 'extracting credentials' do
```
Maybe I should have broken this out, I don't know what it was doing here.

Magisus replied:
Looks like leftover maybe from the work Tony and I did around OAuth tokens. I think we moved the logic (and tests) elsewhere, and probably forgot this.

justinstoller commented on spec/unit/action/deploy/environment_spec.rb:
```diff
@@ -273,16 +282,18 @@
         let(:purge_levels) { [:environment] }
```
Locally, these two tests regarding environment level purging are failing.

justinstoller commented on spec/unit/action/deploy/environment_spec.rb:
```diff
@@ -273,16 +282,18 @@
         end
 
         it "logs that environment was not purged if deploy failed" do
-          expect(subject).to receive(:visit_puppetfile) { subject.instance_variable_set(:@visit_ok, false) }
```
While I'm not sure what is wrong with the first test of these two. This test, I think, is failing because it depends on the instance variable `@visit_ok` being set to false during the deploy. However, I attempted this with no success:
```rb
expect(subject).to receive(:visit_environment).and_wrap_original do |original, env, &block|
  expect(env).to receive(:deploy) { subject.instance_variable_set(:@visit_ok, false) }
  original.call(env)
end
```
without success. (Not sure why I switched the test content back)

I'm wondering if I'm missing something simple in the first test case and if figured out then this one would be fixable by the above code block (or similar)

Magisus replied:
Is this out of date now? All the tests seem to be passing at this point.

justinstoller replied:
Yeah, this is out of date now.

justinstoller commented on spec/unit/action/deploy/environment_spec.rb:
```diff
@@ -275,13 +284,17 @@
         it "only logs about purging environment" do
           expect(subject.logger).to receive(:debug).with(/purging unmanaged content for environment/i)
           expect(subject.logger).to_not receive(:debug).with(/purging unmanaged environments for deployment/i)
-          expect(subject.logger).to_not receive(:debug).with(/purging unmanaged puppetfile content/i)
```
So whenever I tried to implement a similar "wrap_original" expectation on "visit_environment" like above to test for the logger output of "purging unmanaged puppetfile content" the remaining expectations wouldn't fire. I think that's a testing issue though I'm unclear on what the root cause is. I wanted to get other's opinion on it. This test is testing less, though I don't know if it's any worse.

Magisus replied:
Weird that it works in one case but not the other... I feel like I ran into something weird like this recently with some work I was doing on the CA CLI tests, but I can't remember if/how I fixed it... I'll think about it.

Magisus replied:
According to pry, that `env` object doesn't have a `logger` method (or instance var, for that matter). So I _think_ this block is actually throwing and the exception is being squashed, but it's messing with other stuff. I don't know _why_ it doesn't seem to have the method, something to do with weird test doubles maybe?
```
NoMethodError: undefined method `logger' for #<R10K::Environment::Mock:0x00007fcc6bb11220>
```

justinstoller replied:
Ah ha! So it looks like the logger functionality is included in all the child Environment objects, rather than the Environment::Base class. This deployment uses Environment::Mock which doesn't include the logging framework.

Including the R10K::Logging module into the Environment::Mock class causes _a different_ test to fail.... So I think I should move the inclusion of the Logging framework into the Environment::Base and then audit the usages of loggers & the mock environment...

Magisus replied:
That makes sense to me. Yeah, it took me a bit to track down exactly where `logger` was coming from. I expected it to be part of the `Base` environment...

justinstoller replied:
Okay, I moved the logging inclusion for the Environments and while I was at it for the Sources too (since they also have a Mock version that doesn't include a logger). There was also an issue with the test data not being passed down from Source::Mock to Environment::Mock correctly. I think that and the silently failing logger calls were the two issues conspiring against my tests before.

Magisus commented on spec/unit/action/deploy/module_spec.rb:
```diff
@@ -90,8 +90,8 @@
           end
-          @modules << mod
           expect(environment.puppetfile).to receive(:modules).and_return([mod]).twice
+          @modules << mod
```
Why did this need to move?

justinstoller replied:
:facepalm: It didn't. I was messing with this before block and put it back wrong.

Magisus reviewed (commented):
Definitely feel like this is an improvement, that will only get better if/when we remove the rest of the pattern.

justinstoller committed:
(maint) Add Environment#deploy method

This adds a method `deploy` to the Environment classes that syncs
modules similarly to `accept` but does not take a visitor. It also
includes the Puppetfile loading and clean up that would take place
within a `visit_puppetfile` call.

Magisus commented on spec/unit/action/deploy/environment_spec.rb:
```diff
@@ -261,9 +266,13 @@
 
         it "only logs about purging deployment" do
+          expect(subject).to receive(:visit_environment).and_wrap_original do |original, env, &block|
+            expect(env.logger).to_not receive(:debug).with(/purging unmanaged puppetfile content/i)
```
So you know what's bizarre? I don't think this expectation is firing either. I flipped it to `to receive` and the test still passes. I also added a `expect(true).to eq(false)` to this block and it also didn't fail.

Magisus replied:
However, if I drop the `expect` on `env.logger`, the `expect(true)` does trip...

justinstoller committed:
(maint) Move logging inclusion to Environment::Base

Every child Environment includes the Logging module (except for the
mock environment). Include the Logging module in the Environment base
so it is available to all environments.

justinstoller committed:
(maint) Move Logging inclusion to Source::Base

Previously some Sources included the Logging framework, with the end
result that all child Sources (except for the Mock source) had loggers.
This move the inclusion of the Logging framework into the Source::Base
class so all children have logging and where that comes from is easily
understandable.

Magisus reviewed (approved):
Did you want other review on this?
